oauth_dropins/meetup.py

```diff
@@ -0,0 +1,156 @@
+    token = ndb.StringProperty(required=False)
+    state = ndb.TextProperty(required=False)
+
+
+class StartHandler(handlers.StartHandler):
+    """Starts Meetup.com auth. Requests an auth code and expects a redirect back.
+    """
+    def redirect_url(self, state=None):
+        assert (MEETUP_CLIENT_ID and MEETUP_CLIENT_SECRET), (
+                "Please fill in the meetup_client_id and meetup_client_secret files in "
+                "your app's root directory.")
+
+        csrf_key = MeetupCsrf(state=state).put()
+        return GET_AUTH_CODE_URL % {
+                'client_id': MEETUP_CLIENT_ID,
+                'redirect_uri': urllib.parse.quote_plus(self.to_url(state=state)),
+                'state': '%s|%s' % (state, csrf_key.id()),
+                }
+
+
+class CallbackHandler(handlers.CallbackHandler):
+    """The auth callback. Fetches an access token, stores it, and redirects home.
+    """
+    def get(self):
+      # handle errors
+      error = self.request.get('error')
+      if error:
+          if error == 'access_denied':
+              logging.info('User declined')
+              self.finish(None, state=self.request.get('state'))
+              return
+          else:
+              msg = 'Error: %s: %s' % (error, self.request.get('error_description'))
+              logging.info(msg)
+              raise exc.HTTPBadRequest(msg)
+
+      state = util.get_required_param(self, 'state')
+      # lookup the CSRF token
+      try:
+          csrf_id = int(urllib.parse.unquote_plus(state).split('|')[-1])
+      except (ValueError, TypeError):
+          raise exc.HTTPBadRequest('Invalid state value %r' % state)
+
+      csrf = MeetupCsrf.get_by_id(csrf_id)
+      if not csrf:
+          raise exc.HTTPBadRequest('No CSRF token for id %s' % csrf_id)
+
+      # extract auth code and request access token
+      auth_code = util.get_required_param(self, 'code')
+      data = {
+          'code': auth_code,
+          'client_id': MEETUP_CLIENT_ID,
+          'client_secret': MEETUP_CLIENT_SECRET,
+          'grant_type': 'authorization_code',
+          # redirect_uri here must be the same in the oauth code request!
+          # (the value here doesn't actually matter since it's requested server side.)
+          'redirect_uri': self.request.path_url,
+          }
+      resp = util.requests_post(GET_ACCESS_TOKEN_URL,
+                                data=urllib.parse.urlencode(data)).text
+
+      logging.debug('Access token response: %s', resp)
+
+      resp = urllib.parse.parse_qs(resp)
+
+      error = resp.get('error')
+      if error:
+          msg = 'Error: %s: %s' % (error[0], resp.get('error_description'))
+          logging.info(msg)
+          raise exc.HTTPBadRequest(msg)
+
+      access_token = resp['access_token']
+
+      user = BaseAuth.urlopen_access_token(GET_USER_INFO_URL, access_token).read()
+      logging.debug('User info response: %s', user)
+      user_id = json_loads(user)['id']
```

Comment:
mind storing this `user` JSON as a property in `MeetupAuth`, and then ideally pull out their name and return it in `user_display_name()`? see other silos for examples.